### PR TITLE
Ce/app status sql

### PR DESCRIPTION
**Pull request by calellowitz**
@sravfeyn @gcapalbo 

I took a swing at the app status fact sql. It is pretty nasty, but i think it gets the job done. There are likely simpler ways to do the repetitive `CASE` statements with CTEs (https://stackoverflow.com/questions/16959803/select-several-columns-based-on-one-condition) but I think this gets the job done. There are also likely performance improvements that can be made but I am not sure what they are so I did not try that hard to find them. I am happy to revisit any decisions I made along the way, was mostly just looking for something that worked, and did so in one query.

One thing I did do was slightly denormalize the data, i included app_id and user_id, which are the unique keys for this table, but also primary key foreign keys to the app and user dims. it is possible we could use those IDs as the foreign keys if that is seen as better.

### CODE REVIEW

File: corehq/warehouse/models/facts.py [hunk 1]
```diff
@@ -122,27 +116,28 @@ class ApplicationStatusFact(BaseFact, CustomSQLETLMixin):
...
-    # app_dim = models.CharField(max_length=255)
+    app_dim = models.ForeignKey(ApplicationDim, on_delete=models.PROTECT)
+    app_id = models.CharField(max_length=255)
 
-    # TODO: Add domain dimension
-    # domain_dim = models.CharField(max_length=255)
+    domain_dim = models.ForeignKey(DomainDim, on_delete=models.PROTECT)
 
     user_dim = models.ForeignKey(UserDim, on_delete=models.PROTECT)
+    user_id = models.CharField(max_length=255)
```
Comment (gcapalbo) [inline corehq/warehouse/models/facts.py]: i saw your comment about denormalizing and including app_id and user_id but how come that was done - couldn't we just reference those when we make queries by joining to those dims?

**Comment by gcapalbo** [comment]
it's a bit hard to follow what's happening in `app_status_fact.sql`, and i also see it's joining the result sets from two subqueries which won't perform well because there's never indexes to use on those kinds of joins.

i think the sql used here could benefit from being split up into multiple intermediate tables, which would help with organization and with performance.

**Comment by calellowitz** [comment]
@gcapalbo I reorganized a bit, simplified everything up, and got rid of the unnecessary denormalization. Let me know if this looks better.

**Review by snopoke — COMMENTED** [review]
I haven't thought through all the implications of this and I don't want to derail this PR but I was looking at the app status report earlier this week and I think the disjoint between form data and synclog data can make the report very confusing.

I think if we tied the data together with a device ID that would make a lot more logical sense. The report would show data from the most recent 'device, user, app' combo. That way we know that the app version and CC version are consistent. Also the form submission date and sync date would apply to that 'device/app' combo instead of potentially one being from one 'device/app' and the other being from another 'device/app'.

Maybe that adds too much complexity for the edge cases but I think it's worth considering.

Does @dimagi/product have any views about the future / usability of the App Summary report?

Maybe if we renamed it 'Device Summary' it would make more sense.

**Comment by calellowitz** [comment]
Hey @snopoke 
I think you bring up a lot of good points. Having the rows of the app status report be (user, app, device) instead of (user, app) is something I actually suggested in August on an enikshay reports call, in fact there is still a trello card on the enikshay dev board to implement that (https://trello.com/c/JoY29NPM/194-app-status-report-updates-hq). I think that it would definitely reduce certain ambiguities, although not all of them. Specifically the disjoint between sync logs and form submissions could still a happen (a user on the same device could update the app then sync without submitting a new form, leading to a seemingly impossible commcare version/app combo). I believe there were also some uncertainties about webapps at the time.

All of that being said, I was hoping this would be a drop in replacement for the existing report, and potentially behind a feature flag and only on for certain domains, and am a little concerned of the scope creep if we start redesigning the report. From a technical perspective, it should be pretty easy to add device id here, but I am hesitant to add it to the report at this time. Especially as we look at the POC to figure out the viability of this warehouse approach, doing too much other work on the report seems likely to drag this process out even longer

File: corehq/warehouse/models/staging.py [hunk 2]
```diff
@@ -407,3 +398,45 @@ def record_iter(cls, start_datetime, end_datetime):
...
+    user_id = models.CharField(max_length=255, null=True, db_index=True)
+    last_submission = models.DateTimeField(db_index=True)
+    submission_build_version = models.CharField(max_length=255, null=True, db_index=True)
+    commcare_version = models.CharField(max_length=255, null=True, db_index=True)
+
+    @classmethod
+    def dependencies(cls):
+        return [
+            FORM_STAGING_SLUG,
+            APP_STATUS_FACT_SLUG,
```
Comment (gcapalbo) [inline corehq/warehouse/models/staging.py]: how come this is a dependency?

File: corehq/warehouse/models/staging.py [hunk 3]
```diff
@@ -407,3 +398,45 @@ def record_iter(cls, start_datetime, end_datetime):
...
+    last_sync = models.DateTimeField(null=True, db_index=True)
+    domain = models.CharField(max_length=255, null=True, db_index=True)
+    user_id = models.CharField(max_length=255, null=True, db_index=True)
+    sync_build_version = models.CharField(max_length=255, null=True, db_index=True)
+
+    @classmethod
+    def dependencies(cls):
+        return [
+            SYNCLOG_STAGING_SLUG,
+            APP_STATUS_FACT_SLUG,
```
Comment (gcapalbo) [inline corehq/warehouse/models/staging.py]: this one too

File: corehq/warehouse/transforms/sql/app_status_synclog_staging.sql [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    domain,
+    '{{ batch_id }}'
+FROM
+(
+SELECT
+	row_number() over wnd,
+	CASE
+	    WHEN sync_date < app_status.last_sync_log_date
+	    THEN app_status.last_sync_log_date
+	    ELSE first_value(sync_date AT TIME ZONE 'UTC') OVER wnd
```
Comment (gcapalbo) [inline corehq/warehouse/transforms/sql/app_status_synclog_staging.sql]: it looks like this will get the oldest sync date. is that's what's desired? (just asking because it goes into a column called "last_sync")

File: corehq/warehouse/transforms/sql/app_status_synclog_staging.sql [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	END AS last_sync,
+	CASE
+	WHEN sync_date < app_status.last_sync_log_date
+	THEN app_status.last_sync_log_app_build_version
+        ELSE app_dim.version::VARCHAR
+	END as sync_build_version,
+	sync_staging.user_id as user_id,
+	sync_staging.domain as domain
+        FROM {{ synclog_staging }} as sync_staging
+	LEFT JOIN {{ app_status_fact }} as app_status
```
Comment (gcapalbo) [inline corehq/warehouse/transforms/sql/app_status_synclog_staging.sql]: i think i see why app status fact is listed as a dependency above now. just fyi that it creates a circular dependency in the dependency list

File: corehq/warehouse/models/facts.py [hunk 6]
```diff
@@ -122,34 +116,30 @@ class ApplicationStatusFact(BaseFact, CustomSQLETLMixin):
 
     Grain: app_id, user_id
     '''
-    # TODO: Write Update SQL Query (currently there exists a placeholder)
     slug = APP_STATUS_FACT_SLUG
 
-    # TODO: Add app dimension
-    # app_dim = models.CharField(max_length=255)
+    app_id = models.CharField(max_length=255, null=True, db_index=True)
```
Comment (gcapalbo) [inline corehq/warehouse/models/facts.py]: how come this and other fields on the final fact table are nullable?

File: corehq/warehouse/transforms/sql/app_status_form_staging.sql [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    domain,
+    '{{ batch_id }}'
+FROM
+(
+SELECT
+	row_number() over wnd,
+	CASE
+	    WHEN received_on < app_status.last_form_submission_date
+	    THEN app_status.last_form_submission_date
+	    ELSE  first_value(received_on AT TIME ZONE 'UTC') OVER wnd
```
Comment (gcapalbo) [inline corehq/warehouse/transforms/sql/app_status_form_staging.sql]: (same question here, the first received_on would be the oldest one since we're ordering by it)

File: corehq/warehouse/transforms/sql/app_status_synclog_staging.sql (hunk 4 above)
Comment (gcapalbo) [reply corehq/warehouse/transforms/sql/app_status_synclog_staging.sql]: sorry nevermind. i see it's ordering descending

File: corehq/warehouse/transforms/sql/app_status_form_staging.sql (hunk 7 above)
Comment (gcapalbo) [reply corehq/warehouse/transforms/sql/app_status_form_staging.sql]: sorry nevermind. i see it's ordering descending

Comment (gcapalbo) [reply corehq/warehouse/transforms/sql/app_status_form_staging.sql]: actually it occurs to me that we probably can get rid of `first_value` here since we're only taking the first row in the window later on anyway

File: corehq/warehouse/transforms/sql/app_status_synclog_staging.sql [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+	ON sync_staging.user_id = app_status.user_id
```
Comment (gcapalbo) [inline corehq/warehouse/transforms/sql/app_status_synclog_staging.sql]: does this join need app_id too? since it's part of the grain of app_status_fact

File: corehq/warehouse/transforms/sql/app_status_fact.sql [hunk 9]
```diff
@@ -1,45 +1,37 @@
...
-        )
-     ) as t
-     GROUP BY user_id
-) as sync_table
-
-ON sync_table.user_id = form_table.user_id
+FROM
+    {{ app_status_form_staging }} as form_table
+    LEFT JOIN {{ app_status_synclog_staging }} as sync_table
+    ON form_table.user_id = sync_table.user_id
```
Comment (gcapalbo) [inline corehq/warehouse/transforms/sql/app_status_fact.sql]: is the grain for both of these staging tables just one record per user?

File: corehq/warehouse/models/facts.py [hunk 10]
```diff
@@ -122,34 +116,30 @@ class ApplicationStatusFact(BaseFact, CustomSQLETLMixin):
...
 
-    # TODO: Add domain dimension
-    # domain_dim = models.CharField(max_length=255)
+    domain = models.CharField(max_length=255, null=True, db_index=True)
 
-    user_dim = models.ForeignKey(UserDim, on_delete=models.PROTECT)
+    user_id = models.CharField(max_length=255, null=True, db_index=True)
```
Comment (gcapalbo) [inline corehq/warehouse/models/facts.py]: it looks like user_id and app_id are still denormalized - just curious why that was done? if we ever want to join to the user dim or app dim it will take much longer than if we joined on integer ids

**Comment by gcapalbo** [comment]
thanks @calellowitz this is much easier to follow now and will also perform much better.

you may also want someone more familiar than me with with this report to review the transformations.

File: corehq/warehouse/transforms/sql/app_status_synclog_staging.sql (hunk 5 above)
Comment (calellowitz) [reply corehq/warehouse/transforms/sql/app_status_synclog_staging.sql]: We do not actually use the depencies in hq anywhere except adding the table slugs to the sql template context. In airflow, where we handle the actual dependency management, these tables do not depend on the app status fact since we update these first.

File: corehq/warehouse/transforms/sql/app_status_form_staging.sql (hunk 7 above)
Comment (calellowitz) [reply corehq/warehouse/transforms/sql/app_status_form_staging.sql]: good point, i will get rid of that function call

File: corehq/warehouse/transforms/sql/app_status_synclog_staging.sql (hunk 8 above)
Comment (calellowitz) [reply corehq/warehouse/transforms/sql/app_status_synclog_staging.sql]: yes

File: corehq/warehouse/transforms/sql/app_status_fact.sql (hunk 9 above)
Comment (calellowitz) [reply corehq/warehouse/transforms/sql/app_status_fact.sql]: yep

File: corehq/warehouse/models/facts.py (hunk 10 above)
Comment (calellowitz) [reply corehq/warehouse/models/facts.py]: would it work to index those fields on the dim table, for faster joins. I left it like this because I think the vast majority of queries, if not every one, will filter the app status report on user or app or both, and our existing filters output app_id and user_id for those, so each report view would require either a second query to the dim to get the primary key or a join. this schema allows us to do a single query with no joins and get the info we want. I think the alternative is to rewrite the hq filters for reports that use the warehouse or take the performance hit and do the join. the first seems out of scope at this point, and i was not sure if the second was worth it.

Comment (gcapalbo) [reply corehq/warehouse/models/facts.py]: ok if that's the only use case it's probably fine, but only because this is for a very specialized report and probably won't be used anywhere else. if this were a more reusable fact table it would be more important that all foreign keys were integer ids. i also wouldn't recommend joining this to the user dim or app dim with these ids, if that's ever needed then the dim foreign keys should be used and the join would have to happen before filtering.

File: corehq/warehouse/transforms/sql/app_status_synclog_staging.sql (hunk 8 above)
Comment (gcapalbo) [reply corehq/warehouse/transforms/sql/app_status_synclog_staging.sql]: actually, now that i look at the synclog staging table i'm not sure it's needed. what's the grain on synclog staging?

File: corehq/warehouse/models/facts.py (hunk 10 above)
Comment (calellowitz) [reply corehq/warehouse/models/facts.py]: it turns out I needed more information than expected from the user dim, so i am in fact going back to make these normal foreign keys. sorry about all the misdirection

File: corehq/warehouse/transforms/sql/app_status_synclog_staging.sql [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+	LEFT JOIN {{ application_dim }} as app_dim
+	ON sync_staging.build_id = app_dim.application_id
```
[the remark this answers is not in the document]
Comment (calellowitz) [reply corehq/warehouse/transforms/sql/app_status_synclog_staging.sql]: I was going over this with @snopoke earlier and i am going to remove all build info encoded in synclogs from this report. it is not guaranteed to exist and can cause inconsistencies in the data when it differs from the form ones. that will address this issue (i can just get rid of the join) and the comment above when we join the app status fact

**Comment by calellowitz** [comment]
@gcapalbo i think i got to the last things, but let me know if there is anythign else

**Comment by calellowitz** [comment]
> many of the CASE statements do comparisons to columns from a left join, which could be NULL as a result of the left join. note that the result of any comparison with a NULL value (=, <>, <, etc.) is always FALSE. so you may want to review the CASE statements with that in mind

I think is correct. the FALSE clause always includes the new value from the staging table, while the TRUE clause uses the existing app status value. That was something I looked into earlier.

**Comment by calellowitz** [comment]
@gcapalbo i pushed the other fixes
> this join still looks wrong: https://github.com/dimagi/commcare-hq/pull/19731/files#diff-5cab5c0c2ca938abc0de8429be7ed588R26. if there can be more than 1 record per user on synclog staging and more than 1 record per user on app status fact then you'll end up with a multiplication of records in the output

I think this is ok. I changed the window function so it orders by both the synclog date and the one from the app status table. This will duplicate the synclog row (1 per user) for each app in the app status fact table of that user, but the ordering and window function mean it will still choose the correct data, that at row number 1, to insert into the staging table. Let me know if that still does not seem like it will work

**Review by gcapalbo — APPROVED** [review]
looks like some of the stickler comments might be legitimate:

    F821 undefined name 'django'
    W1618 import missing from __future__ import absolute_import
    E131 continuation line unaligned for hanging indent

**Review by gcapalbo — APPROVED** [review]
